### ISSUE TITLE
Disallow disk buffers with sources that checkpoint

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#7340]
#7246 is another example where disk buffers probably aren't necessary, both the `journald` and `file` sources persist data and offer checkpointing capabilities so there is no need to buffer data on disk again. Instead, users should utilize in-memory buffers and allow back pressure to propagate back to the sources where data is already persisted.

A goal of Vector is to reduce the amount of decisions a user has to make and guide them towards optimal performance. Because of this, I think we should disallow disk buffers if all upstream sources offer checkpointing capabilities. I cannot think of a use case where a user would want disk buffering in this scenario. Because disk buffers reduce performance so drastically (~60%) I'd prefer that this be a boot-time error instead of a warning, which will likely be missed.

Curious what others think.

**Comment by jszwedko** [comment]
I tend to agree. I think even just a warning in the docs on buffers would help but a warning on vector start-up would be even better.

**Comment by binarylogic** [comment]
@bruceg, I'm curious if we should lump this in with the e2e ack'ing work since it contributes to the broader cause. Disk buffers reduce the guarantees with that, right?

**Comment by bruceg** [comment]
Disk buffers will initially reduce the guarantees, since events will ack when they are buffered, though we have a plan to allow for acknowledgements on the other side of a disk buffer as well.

I'm not sure this fits in with the broader end-to-end work, though I can see your point. We should at minimum warn about such a configuration, and there would definitely be grounds to either disabling the buffer or failing at startup. The trick may be verifying the sink with the disk buffer feeds only from the one source (or rather, from only sources that forbid a disk buffer).

**Comment by jszwedko** [comment]
[A user left a comment](https://github.com/timberio/vector/issues/7246#issuecomment-833583352) that they'd like to use the `disk` buffer even with the checkpointable `file` source so maybe a warning would be better than a hard error.
